### DISCUSSION TITLE
left recursive rules

### DISCUSSION THREAD

**Discussion by mattkeanny** [textX/textX#367] [Q&A]
Hi,
I am new to parsing and dsl. I have tried antlr4 for a grammar with boolean operators including left recursive rules and I would like to switch to an all-python parser, but I am not sure which one to start with for dealing optimally with the above: TextX/Arpeggio or parglare? guidance would be appreciated. 

Example of the grammar in Antlr format :
expr
 : expr AND expr
 | expr OR expr
 | '(' expr ')'
 | term ;

**Comment by igordejanovic** [comment]
Hi. textX/Arpeggio are based on a recursive descent parsing and do not support left recursive rules. parglare on the other hand is based on bottom-up LR (and its generalized GLR variant) so it accepts left recursive rules and if GLR is used it accepts full set of context-free grammars.

So, the bottom line would be:
- textX is the easiest to start with for DSL development. It is straight forward to understand/debug and it does a lot of heavy lifting for you. But if you need full control over parsing process and transformations or left recursive grammar rules then it is not a good fit.
- If your language is heavily expression based then your grammar will be nicer if you go with parglare (or some other LR parsing library). But, you will have to mold your grammar in the LR class (resolve possible shift/reduce conflicts).
- If you want full freedom with your grammar and you are willing to accept dealing with possible ambiguous parses at run-time, and performance is not crucial then take a look at some generalized parsing approach (like parglare in GLR mode).

**Reply by mattkeanny** [reply]
It would sound that parglare would be the most suitable library, instead of contorting the grammar to fit the peg parser. Thank you for the answer.
